docs/migration/index.md

```diff
@@ -0,0 +1,381 @@
+  )
+
+  app.FeeGrantKeeper = feegrantkeeper.NewKeeper(appCodec, keys[feegrant.StoreKey], app.AccountKeeper)  // <--
+  // Add app.BaseApp as the last argument to upgradekeeper.NewKeeper
+  app.UpgradeKeeper = upgradekeeper.NewKeeper(skipUpgradeHeights, keys[upgradetypes.StoreKey], appCodec, homePath, app.BaseApp)
+  
+  app.IBCKeeper = ibckeeper.NewKeeper(
+    // Add app.UpgradeKeeper
+    appCodec, keys[ibchost.StoreKey], app.GetSubspace(ibchost.ModuleName), app.StakingKeeper, app.UpgradeKeeper, scopedIBCKeeper,
+  )
+
+  govRouter.AddRoute(govtypes.RouterKey, govtypes.ProposalHandler).
+    //...
+    // Replace NewClientUpdateProposalHandler with NewClientProposalHandler
+    AddRoute(ibchost.RouterKey, ibcclient.NewClientProposalHandler(app.IBCKeeper.ClientKeeper))
+
+  // Replace porttypes with ibcporttypes
+  ibcRouter := ibcporttypes.NewRouter()
+
+  app.mm.SetOrderBeginBlockers(
+    upgradetypes.ModuleName,
+    // Add capabilitytypes.ModuleName,
+    capabilitytypes.ModuleName,
+    minttypes.ModuleName,
+    //...
+    // Add feegrant.ModuleName,
+    feegrant.ModuleName,
+  )
+
+  // Add app.appCodec as an argument to module.NewConfigurator:
+  app.mm.RegisterServices(module.NewConfigurator(app.appCodec, app.MsgServiceRouter(), app.GRPCQueryRouter()))
+  
+  // Replace:
+  // app.SetAnteHandler(
+  // 	ante.NewAnteHandler(
+  // 		app.AccountKeeper, app.BankKeeper, ante.DefaultSigVerificationGasConsumer,
+  // 		encodingConfig.TxConfig.SignModeHandler(),
+  // 	),
+  // )
+
+  // With the following:
+  anteHandler, err := ante.NewAnteHandler(
+    ante.HandlerOptions{
+      AccountKeeper:   app.AccountKeeper,
+      BankKeeper:      app.BankKeeper,
+      SignModeHandler: encodingConfig.TxConfig.SignModeHandler(),
+      FeegrantKeeper:  app.FeeGrantKeeper,
+      SigGasConsumer:  ante.DefaultSigVerificationGasConsumer,
+    },
+  )
+  if err != nil {
+    panic(err)
+  }
+  app.SetAnteHandler(anteHandler)
+
+  // Remove the following:
+  // ctx := app.BaseApp.NewUncachedContext(true, tmproto.Header{})
+  // app.CapabilityKeeper.InitializeAndSeal(ctx)
+}
+
+func (app *App) InitChainer(ctx sdk.Context, req abci.RequestInitChain) abci.ResponseInitChain {
+  var genesisState GenesisState
+  if err := tmjson.Unmarshal(req.AppStateBytes, &genesisState); err != nil {
+    panic(err)
+  }
+  // Add the following:
+  app.UpgradeKeeper.SetModuleVersionMap(ctx, app.mm.GetVersionMap())
+  return app.mm.InitGenesis(ctx, app.appCodec, genesisState)
+}
+
+// Replace Marshaler with Codec
+func (app *App) AppCodec() codec.Codec {
+  return app.appCodec
+}
+
+// Replace BinaryMarshaler with BinaryCodec
+func initParamsKeeper(appCodec codec.BinaryCodec, legacyAmino *codec.LegacyAmino, key, tkey sdk.StoreKey) paramskeeper.Keeper {
+  //...
+}
+```
+
+### app/genesis.go
+
+```go
+// Replace codec.JSONMarshaler with codec.JSONCodec
+func NewDefaultGenesisState(cdc codec.JSONCodec) GenesisState {
+  //...
+}
+```
+
+### testutil/keeper/mars.go
```

Comment:
If it's an IBC module the code is different

```go
package keeper

import (
	"testing"

	"github.com/cosmonaut/test/x/mars/keeper"
	"github.com/cosmonaut/test/x/mars/types"
	"github.com/cosmos/cosmos-sdk/codec"
	codectypes "github.com/cosmos/cosmos-sdk/codec/types"
	"github.com/cosmos/cosmos-sdk/store"
	storetypes "github.com/cosmos/cosmos-sdk/store/types"
	sdk "github.com/cosmos/cosmos-sdk/types"
	capabilitykeeper "github.com/cosmos/cosmos-sdk/x/capability/keeper"
	typesparams "github.com/cosmos/cosmos-sdk/x/params/types"
	ibckeeper "github.com/cosmos/ibc-go/modules/core/keeper"
	"github.com/stretchr/testify/require"
	"github.com/tendermint/tendermint/libs/log"
	tmproto "github.com/tendermint/tendermint/proto/tendermint/types"
	tmdb "github.com/tendermint/tm-db"
)

func MarsKeeper(t testing.TB) (*keeper.Keeper, sdk.Context) {
	logger := log.NewNopLogger()

	storeKey := sdk.NewKVStoreKey(types.StoreKey)
	memStoreKey := storetypes.NewMemoryStoreKey(types.MemStoreKey)

	db := tmdb.NewMemDB()
	stateStore := store.NewCommitMultiStore(db)
	stateStore.MountStoreWithDB(storeKey, sdk.StoreTypeIAVL, db)
	stateStore.MountStoreWithDB(memStoreKey, sdk.StoreTypeMemory, nil)
	require.NoError(t, stateStore.LoadLatestVersion())

	registry := codectypes.NewInterfaceRegistry()
	appCodec := codec.NewProtoCodec(registry)
	capabilityKeeper := capabilitykeeper.NewKeeper(appCodec, storeKey, memStoreKey)

	amino := codec.NewLegacyAmino()
	ss := typesparams.NewSubspace(appCodec,
		amino,
		storeKey,
		memStoreKey,
		"MarsSubSpace",
	)
	IBCKeeper := ibckeeper.NewKeeper(
		appCodec,
		storeKey,
		ss,
		nil,
		nil,
		capabilityKeeper.ScopeToModule("MarsIBCKeeper"),
	)

	k := keeper.NewKeeper(
		codec.NewProtoCodec(registry),
		storeKey,
		memStoreKey,
		IBCKeeper.ChannelKeeper,
		&IBCKeeper.PortKeeper,
		capabilityKeeper.ScopeToModule("MarsScopedKeeper"),
	)

	ctx := sdk.NewContext(stateStore, tmproto.Header{}, false, logger)
	return k, ctx
}

```

Reply:
Added!